x/gov/simulation/operations.go

```diff
@@ -0,0 +1,312 @@
+		}
+
+		return simulation.NewOperationMsg(msg, true, ""), nil, nil
+	}
+}
+
+// SimulateMsgVote generates a MsgVote with random values.
+func SimulateMsgVote(ak types.AccountKeeper, k keeper.Keeper) simulation.Operation {
+	return operationSimulateMsgVote(ak, k, simulation.Account{}, -1)
+}
+
+func operationSimulateMsgVote(ak types.AccountKeeper, k keeper.Keeper,
+	simAccount simulation.Account, proposalIDInt int64) simulation.Operation {
+	return func(r *rand.Rand, app *baseapp.BaseApp, ctx sdk.Context, accs []simulation.Account,
+		chainID string) (simulation.OperationMsg, []simulation.FutureOperation, error) {
+		if simAccount.Equals(simulation.Account{}) {
+			simAccount, _ = simulation.RandomAcc(r, accs)
+		}
+
+		var proposalID uint64
+
+		switch {
+		case proposalIDInt < 0:
+			var ok bool
+			proposalID, ok = randomProposalID(r, k, ctx, types.StatusVotingPeriod)
+			if !ok {
+				return simulation.NoOpMsg(types.ModuleName), nil, nil
+			}
+		default:
+			proposalID = uint64(proposalIDInt)
+		}
+
+		option := randomVotingOption(r)
+
+		msg := types.NewMsgVote(simAccount.Address, proposalID, option)
+
+		account := ak.GetAccount(ctx, simAccount.Address)
+		fees, err := simulation.RandomFees(r, ctx, account.SpendableCoins(ctx.BlockTime()))
+		if err != nil {
+			return simulation.NoOpMsg(types.ModuleName), nil, err
+		}
+
+		tx := helpers.GenTx(
+			[]sdk.Msg{msg},
+			fees,
+			chainID,
+			[]uint64{account.GetAccountNumber()},
+			[]uint64{account.GetSequence()},
+			[]crypto.PrivKey{simAccount.PrivKey}...,
+		)
+
+		res := app.Deliver(tx)
+		if !res.IsOK() {
+			return simulation.NoOpMsg(types.ModuleName), nil, errors.New(res.Log)
+		}
+
+		return simulation.NewOperationMsg(msg, true, ""), nil, nil
+	}
+}
+
+// Pick a random deposit
+func randomDeposit(r *rand.Rand, ctx sdk.Context,
+	ak types.AccountKeeper, k keeper.Keeper, addr sdk.AccAddress,
+) (deposit sdk.Coins, skip bool, err error) {
+	account := ak.GetAccount(ctx, addr)
+	coins := account.SpendableCoins(ctx.BlockHeader().Time)
+	if coins.Empty() {
+		return nil, true, nil // skip
+	}
+
+	minDeposit := k.GetDepositParams(ctx).MinDeposit
+	denomIndex := r.Intn(len(minDeposit))
+	denom := minDeposit[denomIndex].Denom
+
+	depositCoins := coins.AmountOf(denom)
+	if depositCoins.IsZero() {
+		return nil, true, nil
+	}
+
+	maxAmt := depositCoins
+	if maxAmt.GT(minDeposit[0].Amount) {
+		maxAmt = minDeposit[0].Amount
+	}
+
+	amount, err := simulation.RandPositiveInt(r, maxAmt)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return sdk.Coins{sdk.NewCoin(denom, amount)}, false, nil
+}
+
+// Pick a random proposal ID from a proposal with a given status.
+// It does not provide a default proposal ID.
+func randomProposalID(r *rand.Rand, k keeper.Keeper,
+	ctx sdk.Context, status types.ProposalStatus) (proposalID uint64, found bool) {
+	proposalID, _ = k.GetProposalID(ctx)
+	if minProposalID == defaultMinProposalID {
```

Comment:
This is a bit difficult to grasp as we're updating a private global variable...which seems like a code smell. Why are we doing this? Can we think of a cleaner approach? In addition, this function needs better documentation.

Reply:
I cleaned it a bit and added a few more comments, Although the cleanest approach would be to get the initial Proposal ID (minProposalID) directly. This is not possible as the state `ProposalID` is updated every time a new proposal is submitted. 

Ideally, `InitialProposalID` should be defined in the gov parameters in order to fix this issue to avoid relying on a global var